core/lib/node/src/node/mod.rs

```diff
@@ -0,0 +1,207 @@
+use std::{collections::HashMap, fmt};
+
+use futures::{future::BoxFuture, FutureExt};
+use tokio::{runtime::Runtime, sync::watch};
+
+pub use self::{context::NodeContext, stop_receiver::StopReceiver};
+use crate::{
+    resource::{ResourceId, ResourceProvider, StoredResource},
+    task::{IntoZkSyncTask, TaskInitError},
+};
+
+mod context;
+mod stop_receiver;
+
+/// "Manager" class of the node. Collects all the resources and tasks,
+/// then runs tasks until completion.
+///
+/// Initialization flow:
+/// - Node instance is created with access to the resource provider.
+/// - Task constructors are added to the node. At this step, tasks are not created yet.
+/// - Optionally, a healthcheck task constructor is also added.
+/// - Once the `run` method is invoked, node
+///   - attempts to create every task. If there are no tasks, or at least task
+///     constructor fails, the node will return an error.
+///   - initializes the healthcheck task if it's provided.
+///   - waits for any of the tasks to finish.
+///   - sends stop signal to all the tasks.
+///   - waits for the remaining tasks to finish.
+///   - calls `after_node_shutdown` hook for every task that has provided it.
+///   - returns the result of the task that has finished.
+pub struct ZkSyncNode {
```

Comment:
Do I understand it right that for multi-machine/multi-process deployment we'd be running multiple instances of `ZkSyncNode` that together consitute one logical zkSync Node? I don't see a way around it, but naming makes it a little confusing. Did you plan at all to represent such multi-machine setups within the framework design?

Reply:
Yes and yes. Naming is subject to changes, of course, but a modular (or microservice, if you must) approach is certainly considered here.